File: app/modules/account_active_timeframes/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.modules.shared.schemas import DashboardModel

ActiveTimeframeMode = Literal["fixed_weekdays", "random_weekly_days"]
ActiveTimeframeAvailability = Literal["always", "active", "inactive", "missing", "invalid"]
ActiveTimeframeAvailabilityReason = Literal["none", "outside_window", "timeframe_missing", "timeframe_invalid"]


class AccountActiveTimeframeResponse(DashboardModel):
    id: str
    display_name: str
    timezone: str
    start_time: str
    end_time: str
    start_minute: int
    end_minute: int
    mode: ActiveTimeframeMode
    weekdays: list[int] = Field(default_factory=list)
    random_days_per_week: int | None = None
    current_weekdays: list[int] = Field(default_factory=list)
    availability: ActiveTimeframeAvailability
    availability_reason: ActiveTimeframeAvailabilityReason
    next_change_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class AccountActiveTimeframesResponse(DashboardModel):
    timeframes: list[AccountActiveTimeframeResponse] = Field(default_factory=list)


class AccountActiveTimeframeUpsertRequest(DashboardModel):
    display_name: str = Field(min_length=1, max_length=255)
    timezone: str = Field(min_length=1)
    start_time: str = Field(min_length=1)
    end_time: str = Field(min_length=1)
    mode: ActiveTimeframeMode
    weekdays: list[int] | None = None
    random_days_per_week: int | None = None


class AccountActiveTimeframeDeleteResponse(DashboardModel):
    status: str


class AccountActiveTimeframeCoverageAccountResponse(DashboardModel):
    account_id: str
    label: str


class AccountActiveTimeframeCoverageSegmentResponse(DashboardModel):
    start: datetime
    end: datetime
    active_account_count: int
    accounts: list[AccountActiveTimeframeCoverageAccountResponse] = Field(default_factory=list)


class AccountActiveTimeframeCoverageSummaryResponse(DashboardModel):
    minimum_coverage: int
    uncovered_minutes: int
    peak_coverage: int
    average_coverage: float
    next_gap_start: datetime | None = None
    next_gap_end: datetime | None = None


class AccountActiveTimeframeCoverageResponse(DashboardModel):
    week_start: datetime
    week_end: datetime
    segments: list[AccountActiveTimeframeCoverageSegmentResponse] = Field(default_factory=list)
    summary: AccountActiveTimeframeCoverageSummaryResponse

File: frontend/src/features/active-timeframes/schemas.ts
import { z } from "zod";

export const ActiveTimeframeModeSchema = z.enum(["fixed_weekdays", "random_weekly_days"]);

export const AccountActiveTimeframeSchema = z.object({
  id: z.string(),
  displayName: z.string(),
  timezone: z.string(),
  startTime: z.string(),
  endTime: z.string(),
  startMinute: z.number().int(),
  endMinute: z.number().int(),
  mode: ActiveTimeframeModeSchema,
  weekdays: z.array(z.number().int()).default([]),
  randomDaysPerWeek: z.number().int().nullable().optional(),
  currentWeekdays: z.array(z.number().int()).default([]),
  availability: z.enum(["always", "active", "inactive", "missing", "invalid"]).or(z.string()),
  availabilityReason: z.string(),
  nextChangeAt: z.string().datetime({ offset: true }).nullable().optional(),
  createdAt: z.string().datetime({ offset: true }),
  updatedAt: z.string().datetime({ offset: true }),
});

export const AccountActiveTimeframesResponseSchema = z.object({
  timeframes: z.array(AccountActiveTimeframeSchema),
});

export const AccountActiveTimeframeUpsertRequestSchema = z.object({
  displayName: z.string().min(1),
  timezone: z.string().min(1),
  startTime: z.string().min(1),
  endTime: z.string().min(1),
  mode: ActiveTimeframeModeSchema,
  weekdays: z.array(z.number().int()).optional(),
  randomDaysPerWeek: z.number().int().nullable().optional(),
});

export const AccountActiveTimeframeDeleteResponseSchema = z.object({
  status: z.string(),
});

export const AccountActiveTimeframeCoverageAccountSchema = z.object({
  accountId: z.string(),
  label: z.string(),
});

export const AccountActiveTimeframeCoverageSegmentSchema = z.object({
  start: z.string().datetime({ offset: true }),
  end: z.string().datetime({ offset: true }),
  activeAccountCount: z.number().int(),
  accounts: z.array(AccountActiveTimeframeCoverageAccountSchema).default([]),
});

export const AccountActiveTimeframeCoverageSummarySchema = z.object({
  minimumCoverage: z.number().int(),
  uncoveredMinutes: z.number().int(),
  peakCoverage: z.number().int(),
  averageCoverage: z.number(),
  nextGapStart: z.string().datetime({ offset: true }).nullable(),
  nextGapEnd: z.string().datetime({ offset: true }).nullable(),
});

export const AccountActiveTimeframeCoverageSchema = z.object({
  weekStart: z.string().datetime({ offset: true }),
  weekEnd: z.string().datetime({ offset: true }),
  segments: z.array(AccountActiveTimeframeCoverageSegmentSchema),
  summary: AccountActiveTimeframeCoverageSummarySchema,
});

export type AccountActiveTimeframe = z.infer<typeof AccountActiveTimeframeSchema>;
export type AccountActiveTimeframeUpsertRequest = z.infer<typeof AccountActiveTimeframeUpsertRequestSchema>;
export type AccountActiveTimeframeCoverage = z.infer<typeof AccountActiveTimeframeCoverageSchema>;
export type AccountActiveTimeframeCoverageSegment = z.infer<typeof AccountActiveTimeframeCoverageSegmentSchema>;
export type AccountActiveTimeframeCoverageSummary = z.infer<typeof AccountActiveTimeframeCoverageSummarySchema>;
